### PR TITLE
go-owasm: Compile wat to wasm

### LINKED ISSUE #1863: go-owasm should expose a function to compile wat to wasm

### PR DESCRIPTION
**Pull request by perimeko**
fixed #1863

### CODE REVIEW

File: go-owasm/api/lib.go [hunk 1]
```diff
@@ -58,3 +60,12 @@ func run(code []byte, isPrepare bool, env EnvInterface) int32 {
 		},
 	}))
 }
+
+func WatToWasm(code []byte) ([]byte, error) {
```
Comment (sorawit) [inline go-owasm/api/lib.go]: Let's call this`Wat2Wasm` for the sake of consistency

File: go-owasm/api/lib.go [hunk 2]
```diff
@@ -58,3 +60,12 @@ func run(code []byte, isPrepare bool, env EnvInterface) int32 {
...
+	inputSpan := copySpan(code)
+	defer freeSpan(inputSpan)
+	outputSpan := newSpan(1 * 1024 * 1024)
```
Comment (sorawit) [inline go-owasm/api/lib.go]: Eventually we will need to take this as parameter or define it as a constant somewhere.

File: go-owasm/src/error.rs [hunk 3]
```diff
@@ -0,0 +1,11 @@
+#[repr(i32)]
+pub enum Error {
+  Ok = 0,
+  CompileFail = 1,
```
Comment (sorawit) [inline go-owasm/src/error.rs]: Please remove unused errors.

File: go-owasm/src/error.rs [hunk 4]
```diff
@@ -0,0 +1,11 @@
+#[repr(i32)]
+pub enum Error {
+  Ok = 0,
```
Comment (sorawit) [inline go-owasm/src/error.rs]: Let's suffix them with "Error"
```
NoError = 0
CompliationError = 1
ParseError = 2
...
```

**Review by sorawit — CHANGES_REQUESTED** [review]

File: go-owasm/src/error.rs (hunk 3 above)
Comment (perimeko) [reply go-owasm/src/error.rs]: This error is used in Compile function

File: go-owasm/api/lib_test.go [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+	wasm, err := Wat2Wasm(code)
+	require.NoError(t, err)
+
+	expectedWasm := readWasmFile("test")
+	require.Equal(t, expectedWasm, wasm)
+}
+
+func TestFailEmptyWatContent(t *testing.T) {
+	code := []byte("")
+	wasm, err := Wat2Wasm(code)
```
Comment (taobun) [inline go-owasm/api/lib_test.go]: I think we don't need to test wasm if an error happens.

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/x/oracle/keeper/owasm.go [hunk 6]
```diff
@@ -43,9 +43,9 @@ func (k Keeper) PrepareRequest(ctx sdk.Context, r types.RequestSpec, ibcInfo *ty
 		return err
 	}
 	code := k.GetFile(script.Filename)
-	exitCode := owasm.Prepare(code, env) // TODO: Don't forget about prepare gas!
-	if exitCode != 0 {
-		k.Logger(ctx).Info(fmt.Sprintf("failed to prepare request with code: %d", exitCode))
+	err = owasm.Prepare(code, env) // TODO: Don't forget about prepare gas!
+	if err != nil {
+		k.Logger(ctx).Info(fmt.Sprintf("failed to prepare request with error: %s", err.Error()))
```
Comment (sorawit) [inline chain/x/oracle/keeper/owasm.go]: Let's wrap error and stop logging.

File: chain/x/oracle/keeper/owasm.go [hunk 7]
```diff
@@ -92,11 +92,11 @@ func (k Keeper) ResolveRequest(ctx sdk.Context, reqID types.RequestID) {
...
 	script := k.MustGetOracleScript(ctx, req.OracleScriptID)
 	code := k.GetFile(script.Filename)
-	exitCode := owasm.Execute(code, env) // TODO: Don't forget about gas!
+	err := owasm.Execute(code, env) // TODO: Don't forget about gas!
 	var res types.OracleResponsePacketData
-	if exitCode != 0 {
+	if err != nil {
 		k.Logger(ctx).Info(fmt.Sprintf(
-			"failed to execute request id: %d with code: %d", reqID, exitCode,
+			"failed to execute request id: %d with error: %s", reqID, err.Error(),
```
Comment (sorawit) [inline chain/x/oracle/keeper/owasm.go:98]: ditto.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/oracle/keeper/owasm.go (hunk 7 above)
Comment (taobun) [reply chain/x/oracle/keeper/owasm.go:98]: Use log when executing code.

File: go-owasm/api/lib_test.go [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+func TestFailEmptyWatContent(t *testing.T) {
+	code := []byte("")
+	_, err := Wat2Wasm(code)
+	require.Equal(t, ErrParseError, err)
+}
+
+func TestFailInvalidWatContent(t *testing.T) {
+	code := []byte("invalid wat content")
+	wasm, err := Wat2Wasm(code)
```
Comment (taobun) [inline go-owasm/api/lib_test.go]: ditto

**Review by taobun — CHANGES_REQUESTED** [review]

**Review by sorawit — APPROVED** [review]

**Review by taobun — APPROVED** [review]

**Pull request merged**